contracts/governance/extensions/GovernorStorage.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.19;
+
+import "../Governor.sol";
+
+/**
+ * @dev Extension of {Governor} that implements storage of proposal details. This modules also provides primitives for
+ * the enumerability of proposals.
+ *
+ * Usecases for this module include:
+ * - UIs that expore the proposal state without relying on event indexing.
+ * - For some L2 chains where storage is compared to calldata, using a the proposalId version of queue and execute
+ *   might me cheaper.
+ *
+ * _Available since v5.0._
+ */
+abstract contract GovernorStorage is Governor {
+    struct ProposalDetails {
+        address[] targets;
+        uint256[] values;
+        bytes[] calldatas;
+        bytes32 descriptionHash;
+    }
+
+    uint256[] private _proposalIds;
+    mapping(uint256 => ProposalDetails) private _proposalDetails;
+
+    /**
+     * @dev See {IGovernor-propose}.
+     */
+    function propose(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        string memory description
+    ) public virtual override returns (uint256) {
+        uint256 proposalId = super.propose(targets, values, calldatas, description);
+        // store
+        _proposalIds.push(proposalId);
+        _proposalDetails[proposalId] = ProposalDetails({
+            targets: targets,
+            values: values,
+            calldatas: calldatas,
+            descriptionHash: keccak256(bytes(description))
```

Comment:
I'm concerned about this because we're double hashing (expensive operation) in `super.propose()` and also here.
Is it possible to split `Governor.sol`'s `propose` into an internal `_propose` that already receives a `descriptionHash`? This way we can reuse it without hashing twice

Reply:
I fell like splitting `propose` would add complexity ... for something that is not a big gain.

Hashing is not the expensive part here:

- Doing keccak for ~1000 bytes is <500 gas. 
- This operation writes about `4+3*l` slots, with `l` the "length" or the proposal ... and more if calldatas are multi-slot (ofter the case).
  - These slots are all expected to be empty, so that is 21k per slot
    - so that is `84k + 63k * length` 
Even if the proposal contains a single call, and that call has very little calldata, we are still way over 100k gas, so I wouldn't worry about saving 500.

Again, this function is intended to be used on chain where gas is cheap.

Reply:
Ok, I agree, seems like I overestimated what the double operation would be.